src/server/pkg/storage/track/gc.go

```diff
@@ -0,0 +1,93 @@
+package track
+
+import (
+	"context"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type Deleter interface {
+	Delete(ctx context.Context, id string) error
+}
+
+type DeleterMux func(string) Deleter
+
+func (dm DeleterMux) Delete(ctx context.Context, id string) error {
+	deleter := dm(id)
+	if deleter == nil {
+		return errors.Errorf("deleter mux does not have deleter for (%s)", id)
+	}
+	return deleter.Delete(ctx, id)
+}
+
+type GarbageCollector struct {
+	tracker Tracker
+	period  time.Duration
+	deleter Deleter
+}
+
+func NewGarbageCollector(tracker Tracker, period time.Duration, deleter Deleter) *GarbageCollector {
+	return &GarbageCollector{
+		tracker: tracker,
+		period:  period,
+		deleter: deleter,
+	}
+}
+
+func (gc *GarbageCollector) Run(ctx context.Context) error {
+	ticker := time.NewTicker(gc.period)
+	defer ticker.Stop()
+	for {
+		ctx, cf := context.WithTimeout(ctx, gc.period/2)
```

Comment:
It would be good to follow the same pattern as the renewer where you have a wrapper function so you can defer the cancel function and just return from runUntilEmpty. Also, we discussed this offline, but lets have this always return an error if it returns, which I think should help us address the messiness in the pfs master. We also might want a retry in this code path somewhere, I am guessing where we do the `IterateDeletable`.

Reply:
The GC loop will log any errors and the ticker keeps going, so I don't think we need another retry loop right now.